src/search/parser/abstract_syntax_tree.cc

```diff
@@ -0,0 +1,489 @@
+        variables.erase(name);
+    }
+
+    bool has_variable(const string &name) const {
+        return variables.count(name);
+    }
+
+    const plugins::Type &get_variable_type(const string &name) {
+        assert(has_variable(name));
+        return *variables[name];
+    }
+
+    const plugins::Registry &get_registry() const {return registry;}
+};
+
+template<typename T, typename K>
+static vector<T> get_keys(const unordered_map<T, K> &map) {
+    vector<T> keys;
+    keys.reserve(map.size());
+    for (const auto &key_value : map) {
+        keys.push_back(key_value.first);
+    }
+    return keys;
+}
+
+static ASTNodePtr parse_ast_node(const string &definition, DecorateContext &context) {
+    ASTNodePtr node = nullptr;
+    try {
+        TokenStream tokens = split_tokens(definition);
+        node = parse(tokens);
+    }  catch (ParserError &e) {
+        context.parser_error(e.get_message());
+    }
+    return node;
+}
+
+DecoratedASTNodePtr ASTNode::decorate() const {
+    DecorateContext context;
+    context.push_layer("Start semantic analysis");
+    return decorate(context);
+}
+
+LetNode::LetNode(const string &variable_name, ASTNodePtr variable_definition,
+                 ASTNodePtr nested_value)
+    : variable_name(variable_name), variable_definition(move(variable_definition)),
+      nested_value(move(nested_value)) {
+}
+
+DecoratedASTNodePtr LetNode::decorate(DecorateContext &context) const {
+    context.push_layer("Let: " + variable_name);
+    if (context.has_variable(variable_name)) {
+        context.parser_error("Variable is already defined in the current scope.");
+    }
+    const plugins::Type &var_type = variable_definition->get_type(context);
+    if (!var_type.supports_variable_binding()) {
+        context.parser_error(
+            "The value of variable '" + variable_name +
+            "' is not permitted to be assigned to a variable.");
+    }
+
+    context.push_layer("Check variable definition.");
+    DecoratedASTNodePtr decorated_definition =
+        variable_definition->decorate(context);
+    context.pop_layer();
+
+    context.push_layer("Check expression.");
+    context.add_variable(variable_name, var_type);
+    DecoratedASTNodePtr decorated_nested_value = nested_value->decorate(context);
+    context.remove_variable(variable_name);
+    context.pop_layer();
+
+    context.pop_layer();
+    return utils::make_unique_ptr<DecoratedLetNode>(
+        variable_name, move(decorated_definition), move(decorated_nested_value));
+}
+
+void LetNode::dump(string indent) const {
+    cout << indent << "LET:" << variable_name << " = " << endl;
+    indent = "| " + indent;
+    variable_definition->dump(indent);
+    cout << indent << "IN:" << endl;
+    nested_value->dump("| " + indent);
+}
+
+const plugins::Type &LetNode::get_type(DecorateContext &context) const {
+    const plugins::Type &variable_type = variable_definition->get_type(context);
+    context.add_variable(variable_name, variable_type);
```

Comment:
The type of an expression depends on what variables are defined and what type they have. For example, `h` has type `Symbol` if it occurs just by itself, but in this context `let(h, ff(), h)`, the type of the second `h` is `Evaluator` because it is a variable bound to an evaluator.